pkg/liveshare/client.go

```diff
@@ -112,15 +119,49 @@ func Connect(ctx context.Context, opts Options) (*Session, error) {
 	s := &Session{
 		ssh:             ssh,
 		rpc:             rpc,
+		grpc:            grpc.New(),
 		clientName:      opts.ClientName,
 		keepAliveReason: make(chan string, 1),
 		logger:          opts.Logger,
 	}
 	go s.heartbeat(ctx, 1*time.Minute)
 
+	// Connect to the gRPC server so we can make requests anywhere we have access to the session
+	s.connectToGrpcServer(ctx, opts.SessionToken)
+
 	return s, nil
 }
 
+// Connects to the gRPC server running on the host VM
+func (s *Session) connectToGrpcServer(ctx context.Context, token string) error {
+	listen, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", 0))
+	if err != nil {
+		return err
+	}
+
+	// Tunnel the remote gRPC server port to the local port
+	localGrpcServerPort := listen.Addr().(*net.TCPAddr).Port
+	internalTunnelClosed := make(chan error, 1)
+	go func() {
+		fwd := NewPortForwarder(s, codespacesInternalSessionName, codespacesInternalPort, true)
+		internalTunnelClosed <- fwd.ForwardToListener(ctx, listen)
+	}()
+
+	// Make a connection to the gRPC server
+	err = s.grpc.Connect(ctx, localGrpcServerPort, token)
```

Comment:
note: just as a general observation, there's an inherent race between these two processes. 

1. A listener starts listening on some port (good)
2. A go routine is launched to handle connections on that listener
3. Concurrently, we have the grpc client connect to the listener with a timeout of 30 seconds

I believe that usually most of the times, the port forwarder will be ready within 30 seconds and thus the grpc client won't timeout but it is worth calling out that this is racy. Not sure if retries might be worth thinking about.

Another thing to think about is that 30 seconds is quite a lot of time, I am not sure of any other operation in the CLI where we give it such a long time to finish, perhaps considering lowering this timeout.

Reply:
>What's different between this and doing the same [here](https://github.com/dmgardiner25/cli/blob/3f34d6d95e6e5e5a914dac13dcf0829a9fd2fce1/pkg/cmd/codespace/jupyter.go#L61) to forward the JupyterLab server port? Could we not also run into the same issue there?

Great question, that is racy as well, but a little different since the browser has an insanely long timeout of 5 minutes. 

This is why my comment is more of an observation of something to be aware of.